Lib/imath.py

```diff
@@ -0,0 +1,196 @@
+        for a in bases
+    )
+
+
+def is_prime(n: int, s: int=25) -> bool:
+    """Returns whether n is prime.
+
+    A deterministic version of the Miller-Rabin primality test is used for smalls
+    n. For large n, a probabilistic Miller-Rabin test is used and n is either
+    known to be composite when is_prime(n) returns True, or probably prime with
+    with a probability of failure of at most 4^(−s).
+
+    >>> is_prime(17)
+    True
+    >>> is_prime(121)
+    False
+    """
+
+    if n <= 1:
+        return False
+    if n%2 == 0:
+        return (n == 2)
+
+    # If n is small with test we use a deterministic Miller-Rabin test,
+    # see https://miller-rabin.appspot.com/ and
+    # https://arxiv.org/pdf/1509.00864.pdf
+    if n < 341531:
+        bases = [9345883071009581737]
+    elif n < 1050535501:
+        bases = [336781006125, 9639812373923155]
+    elif n < 350269456337:
+        bases = [4230279247111683200, 14694767155120705706, 16641139526367750375]
+    elif n < 55245642489451:
+        bases = [2, 141889084524735, 1199124725622454117, 11096072698276303650]
+    elif n < 7999252175582851:
+        bases = [2, 4130806001517, 149795463772692060, 186635894390467037, 3967304179347715805]
+    elif n < 585226005592931977:
+        bases = [2, 123635709730000, 9233062284813009, 43835965440333360, 761179012939631437, 1263739024124850375]
+    elif n < 18446744073709551616:
+        bases = [2, 325, 9375, 28178, 450775, 9780504, 1795265022]
+    elif n < 318665857834031151167461:
+        bases = [2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37]
+    elif n < 3317044064679887385961981:
+        bases = [2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41]
+    else:
+        bases = (randrange(2, n) for _ in range(s))
+
+    return _miller_rabin(bases, n)
+
+
+def next_prime(n: int) -> int:
+    """Returns the smallest prime number greater than n.
+
+    >>> next_prime(6)
+    7
+    """
+    if n < 0:
+        n = 0
+    if n < 10:
+        return [2, 2, 3, 5, 5, 7, 7, 11, 11, 11][n]
+
+    if 1 <= n % 6 <= 4:
+        n += 5 - (n%6)
+        if is_prime(n):
+            return n
+
+    if n % 6 == 0:
+        n += 1
+    else:
+        n += 2
+
+    while True:
+        if is_prime(n):
+            return n
+        n += 4
+        if is_prime(n):
+            return n
+        n += 2
+
+
+def previous_prime(n: int) -> int:
+    """Returns the largest prime number less than n.
+
+    >>> previous_prime(23)
+    19
+    """
+    if n <= 2:
+        raise ValueError(f'There is no prime number less than {n}.')
+
+    if n < 10:
+        return [0, 0, 0, 2, 3, 3, 5, 5, 7, 7, 7][n]
+
+    if 1 < n % 6 <= 5:
+        n -= (n%6) - 1
+        if is_prime(n):
+            return n
+
+    if n % 6 == 0:
+        n -= 1
+    else:
+        n -= 2
+
+    while True:
+        if is_prime(n):
+            return n
+        n -= 4
+        if is_prime(n):
+            return n
+        n -= 2
+
+
+def _rho_pollard(n):
```

Comment:
The following should be exactly the same algorithm, just refactored a little bit, and it runs a little faster for a few tests by my measurements.
```python
def _rho_pollard(n):
    # From https://en.wikipedia.org/wiki/Pollard%27s_rho_algorithm
    # and https://en.wikipedia.org/wiki/Cycle_detection#Brent.27s_algorithm
    tortoise = hare = 2
    power = 1
    _gcd = gcd
    _repeat = itertools.repeat
    while True:
        # Double the length of the run each time
        for _ in _repeat(None, power):
            hare = (hare**2 + 1) % n
            d = _gcd(hare-tortoise, n)
            if d == 1:
                continue
            elif d == n:
                # failure; start another
                tortoise = hare = randrange(n)
                power = 1
                break
            else:
                return d
        else: # no break
            tortoise = hare
            power *= 2
```

Changes (feel free to pick and choose as you desire):

* starting at 2 rather than random seems to do good things
* Limit the Python `int` comparisons to one per inner loop
    * Factor out the inner loop as a `repeat` (a little faster than `range()`), making c code do the comparisons
* making _gcd and _repeat local variables helps a little.